physics/n_body_simulation.py

```diff
@@ -0,0 +1,337 @@
+    """
+
+    position_x = 0.9700436
+    position_y = -0.24308753
+    velocity_x = 0.466203685
+    velocity_y = 0.43236573
+
+    bodies1 = [
+        Body(position_x, position_y, velocity_x, velocity_y, size=0.2, color="red"),
+        Body(-position_x, -position_y, velocity_x, velocity_y, size=0.2, color="green"),
+        Body(0, 0, -2 * velocity_x, -2 * velocity_y, size=0.2, color="blue"),
+    ]
+    body_system1 = BodySystem(bodies1, time_factor=3)
+    plot("Figure-8 solution to the 3-body-problem", body_system1, -2, 2, -2, 2)
+
+
+def example_2() -> None:
+    """
+    Example 2: Moon's orbit around the earth
+    This example can be seen as a test of the implementation: given the right
+    initial conditions, the moon should orbit around the earth as it actually does.
+    (mass, velocity and distance taken from https://en.wikipedia.org/wiki/Earth
+    and https://en.wikipedia.org/wiki/Moon)
+    No doctest provided since this function does not have a return value.
+    """
+
+    moon_mass = 7.3476e22
+    earth_mass = 5.972e24
+    velocity_dif = 1022
+    earth_moon_distance = 384399000
+    gravitation_constant = 6.674e-11
+
+    # Calculation of the respective velocities so that total impulse is zero,
+    # i.e. the two bodies together don't move
+    moon_velocity = earth_mass * velocity_dif / (earth_mass + moon_mass)
+    earth_velocity = moon_velocity - velocity_dif
+
+    moon = Body(-earth_moon_distance, 0, 0, moon_velocity, moon_mass, 10000000, "grey")
+    earth = Body(0, 0, 0, earth_velocity, earth_mass, 50000000, "blue")
+    body_system2 = BodySystem([earth, moon], gravitation_constant, time_factor=1000000)
+    plot(
+        "Moon's orbit around the earth",
+        body_system2,
+        -430000000,
+        430000000,
+        -430000000,
+        430000000,
+    )
+
+
+def example_3() -> None:
+    """
+    Example 3: Random system with many bodies.
+    No doctest provided since this function does not have a return value.
+    """
+
+    bodies = []
+    for i in range(10):
+        velocity_x = random.uniform(-0.5, 0.5)
+        velocity_y = random.uniform(-0.5, 0.5)
+
+        # Bodies are created pairwise with opposite velocities so that the
+        # total impulse remains zero
+        bodies.append(
+            Body(
+                random.uniform(-0.5, 0.5),
+                random.uniform(-0.5, 0.5),
+                velocity_x,
+                velocity_y,
+                size=0.05,
+            )
+        )
+        bodies.append(
+            Body(
+                random.uniform(-0.5, 0.5),
+                random.uniform(-0.5, 0.5),
+                -velocity_x,
+                -velocity_y,
+                size=0.05,
+            )
+        )
+    body_system3 = BodySystem(bodies, 0.01, 10, 0.1)
+    plot("Random system with many bodies", body_system3, -1.5, 1.5, -1.5, 1.5)
+
+
+if __name__ == "__main__":
+    example_1()
+    example_2()
+    example_3()

```

Comment:
Let's move the plotting logic outside the example functions and have each example function return a BodySystem so that they can be properly tested...
```suggestion
    plot("Figure-8 solution to the 3-body-problem", example_1(), -2, 2, -2, 2)
    plot(
        "Moon's orbit around the earth",
        example_2(),
        -430000000,
        430000000,
        -430000000,
        430000000,
    )
    plot("Random system with many bodies", example_3(), -1.5, 1.5, -1.5, 1.5)

```